src/track/beats.h

```diff
@@ -54,124 +45,232 @@ class Beats : public QObject {
+    /// curFrameNum.  (An n of 4 results in an averaging of 8 beats).  Invalid
+    /// BPM returns Bpm().
+    Bpm getBpmAroundPosition(FramePos curFrame, int n) const;
+
+    /// Sets the track signature starting at specified bar
+    void setSignature(TimeSignature sig, int downbeatIndex);
+
+    /// Translate all beats in the song by numFrames. Beats that lie
+    /// before the start of the track or after the end of the track are not
+    /// removed.
+    void translate(FrameDiff_t numFrames);
+
+    /// Scale the position of every beat in the song by dScalePercentage.
+    void scale(enum BeatsInternal::BPMScale scale);
+
+    /// Set bpm marker at a beat
+    void setBpm(Bpm bpm, int beatIndex = 0);
+
+    /// Returns the number of beats
+    inline int size() {
+        return m_beatsInternal.size();
     }
 
-    ////////////////////////////////////////////////////////////////////////////
-    // Beat mutations
-    ////////////////////////////////////////////////////////////////////////////
+    /// Returns the frame number for the first beat, -1 is no beats
+    FramePos getFirstBeatPosition() const;
 
-    // Add a beat at location dBeatSample. Beats instance must have the
-    // capability BEATSCAP_ADDREMOVE.
-    virtual void addBeat(double dBeatSample) = 0;
+    /// Returns the frame number for the last beat, -1 if no beats
+    FramePos getLastBeatPosition() const;
 
-    // Remove a beat at location dBeatSample. Beats instance must have the
-    // capability BEATSCAP_ADDREMOVE.
-    virtual void removeBeat(double dBeatSample) = 0;
+    /// Return the sample rate
+    SINT getSampleRate() const;
 
-    // Translate all beats in the song by dNumSamples samples. Beats that lie
-    // before the start of the track or after the end of the track are not
-    // removed. Beats instance must have the capability BEATSCAP_TRANSLATE.
-    virtual void translate(double dNumSamples) = 0;
+    /**
+     * Convert a non-downbeat to a downbeat shifting all downbeats
+     * @param beatIndex Index of the beat to be converted to downbeat
+     */
+    void setAsDownbeat(int beatIndex);
 
-    // Scale the position of every beat in the song by dScalePercentage. Beats
-    // class must have the capability BEATSCAP_SCALE.
-    virtual void scale(enum BPMScale scale) = 0;
+    /// Prints debugging information in stderr
+    friend QDebug operator<<(QDebug dbg, const BeatsPointer& arg);
 
-    // Adjust the beats so the global average BPM matches dBpm. Beats class must
-    // have the capability BEATSCAP_SET.
-    virtual void setBpm(double dBpm) = 0;
+    /**
+     * Set sample rate of the track which owns these beats.
+     * @param sampleRate Track sample rate
+     */
+    void setSampleRate(int sampleRate);
 
-    virtual SINT getSampleRate() const = 0;
+    /**
+     * Set duration of the track which owns these beats.
+     * @param duration Track duration in seconds.
```

Comment:
Instead of adding a comment you could name the parameter `trackDurationInSeconds`.

Reply:
Or the function itself, i.e. `setTrackDurationInSeconds` to be more specific.

Reply:
We might also store the whole `StreamInfo` in BeatsInternal instead of individual values:

```
std::optional<mixxx::audio::SteamInfo> m_streamInfo;
```

- it could be missing/unknown
- it is injected from outside and doesn't belong to the other properties that are stored in the protobuf, i.e. it is a *transient* property

Reply:
It would be even favorable to keep this `m_streamInfo` member in `Beats` and instead pass the current value of selected properties as additional function parameters to `BeatsInternal` only when needed.

Reply:
Since all properties of `StreamInfo` have (invalid) default values we could try to omit wrapping it into `std::optional` to not complicate the implementation.